File: prop-archaius/src/main/java/io/pleo/prop/archaius/ArchaiusProp.java
package io.pleo.prop.archaius;

import java.time.Instant;

import com.netflix.config.Property;

import io.pleo.prop.core.Prop;

/**
 * Prop that wraps a com.netflix.config.Property
 */
public class ArchaiusProp<T> implements Prop<T> {
  private Property<T> archaiusProperty;

  public ArchaiusProp(Property<T> archaiusProperty) {
    this.archaiusProperty = archaiusProperty;
  }

  @Override
  public T get() {
    return archaiusProperty.getValue();
  }

  @Override
  public T invoke() {
    return get();
  }

  @Override
  public String getName() {
    return archaiusProperty.getName();
  }

  @Override
  public Instant getChangedTimestamp() {
    return Instant.ofEpochMilli(archaiusProperty.getChangedTimestamp());
  }

  @Override
  public void addCallback(Runnable callback) {
    archaiusProperty.addCallback(callback);
  }

  @Override
  public void removeAllCallbacks() {
    archaiusProperty.removeAllCallbacks();
  }

  @Override
  public String toString() {
    return archaiusProperty.toString();
  }
}

File: prop-archaius/src/main/kotlin/io/pleo/prop/archaius/ArchaiusProp.kt
package io.pleo.prop.archaius

import com.netflix.config.Property
import io.pleo.prop.core.Callback
import io.pleo.prop.core.Prop
import java.time.Instant

/**
 * Prop that wraps a com.netflix.config.Property
 */
class ArchaiusProp<T>(private val archaiusProperty: Property<T>) : Prop<T> {
    override fun get(): T {
        return archaiusProperty.value
    }

    override fun invoke(): T {
        return get()
    }

    override val name: String
        get() = archaiusProperty.name

    override val changedTimestamp: Instant
        get() = Instant.ofEpochMilli(archaiusProperty.changedTimestamp)

    override fun addCallback(callback: Callback<T>) =
        archaiusProperty.addCallback {
            callback(get())
        }

    override fun removeAllCallbacks() = archaiusProperty.removeAllCallbacks()
    override fun toString() = archaiusProperty.toString()
}
